talisker/wsgi.py

```diff
@@ -276,20 +275,33 @@ def __next__(self):
         """
         if self.iter is None:
             raise Exception("WSGIResponse: iterator has not been set yet")
-        # We don't actually call the provided start_response until we start
-        # iterating the content. This provides us with more control over the
-        # response, and allows us to more cleanly switch to an error response
-        # regardless of WSGI server implementation. Talisker apps can call
-        # start_response multiple times, and only the final call will influence
-        # the response status and headers. In Gunicorn for example, the headers
-        # from all calls to start_response would be sent, which usually not
-        # correct, and leads to duplictation or conflict of headers
+        # We don't actually call the WSGI server's provided start_response
+        # until we are ready to start iterating the content. This provides us
+        # with more control over the response, and allows us to more cleanly
+        # switch to an error response regardless of WSGI server implementation.
+        # Talisker apps can call start_response multiple times, and only the
+        # final call will influence the response status and headers. In
+        # Gunicorn for example, the headers from all calls to start_response
+        # would be sent, which usually not correct, and leads to duplictation
+        # or conflict of headers
+
+        # There is some finesse here around calling start_response(), to
+        # support different behaviours from our wrapped WSGI app. Most WSGI
+        # apps call start_response() before returning their iterator, and
+        # indeed must do that if they have an empty iterator.
+        #
+        # Other apps can be lazy, and do not call start_response() until their
+        # iterator is started (e.g. werkzeug's DebuggedApplication).
+        # So we try calling start_response() just *before* iteration, as well
+        # as also after the first iteration, to work with both models.
+        if self.status is not None and not self.start_response_called:
+            self.call_start_response()
+
         try:
             chunk = next(self.iter)
-            # some WSGI apps don't actually call start_response() until they've
-            # started iteration, so we delay calling the WSGI server's start
-            # response until the last possible moment
-            self.ensure_start_response()
+            # support lazy WSGI apps, as above
```

Comment:
shouldn't this be in an `else:` clause - we don't expect `start_response()` to call `StopIteration` (or `GeneratorExit` or ... any of the others?)

Reply:
Hmm, calling the upstream start_response can raise, if the headers have already been sent, so my think it's better to keep it in the try: block, as that way we'd get a sentry report if it raises.